ncm-cron/src/main/pan/components/cron/schema.pan

```diff
@@ -27,40 +27,123 @@ type structure_cron_syslog = {
+    lower_bound = ARGV[1];
+    upper_bound = ARGV[2];
+    text_regex = ARGV[3];
+
+    # Check that the field contains only valid characters
+    if (!match(timing_value, '^(?:[a-z0-9/*-,]+)$')) error(format('"%s" contains invalid characters', timing_value));
+
+    # Find runs of digits and validate them against provided bounds
+    foreach(k; v; matches(timing_value, '([0-9]+)')) {
+        i = to_long(v);
+        if (i < lower_bound) error(format('Value %d is below lower bound of %d', i, lower_bound));
+        if (i > upper_bound) error(format('Value %d is above upper bound of %d', i, upper_bound));
+    };
+
+    # Find runs of letters and validate them against provided regex
+    foreach(k; v; matches(timing_value, '([a-z]+)')) {
+        if (!match(v, text_regex)) error(format('"%s" is not a valid value for this field item', v));
+    };
+
+    # Ignore all other characters
+    return(true);
+};
+
+@documentation{ Convenience wrapper for validating cron minute field }
+function valid_cron_minute = valid_cron_timing(ARGV[0], 0, 59, '^(?![a-z]).+$');
+
+@documentation{ Convenience wrapper for validating cron hour field }
+function valid_cron_hour = valid_cron_timing(ARGV[0], 0, 23, '^(?![a-z]).+$');
+
+@documentation{ Convenience wrapper for validating cron day of month field }
+function valid_cron_day_of_month = valid_cron_timing(ARGV[0], 1, 31, '^(?![a-z]).+$');
+
+@documentation{ Convenience wrapper for validating cron month field }
+function valid_cron_month = valid_cron_timing(ARGV[0], 1, 12, '^(?:jan|feb|mar|apr|may|jun|jul|aug|sep|oct|nov|dec)$');
+
+@documentation{ Convenience wrapper for validating cron day of week field }
+function valid_cron_day_of_week = valid_cron_timing(ARGV[0], 0, 7, '^(?:mon|tue|wed|thu|fri|sat|sun)$');
+
 type structure_cron_timing = {
-    'minute' ? string
-    'hour' ? string
-    'day' ? string
-    'month' ? string
-    'weekday' ? string
+    @documentation{ minute of hour (0-59) }
+    'minute' ? string with valid_cron_minute(SELF)
+    @documentation{ hour of day (0-23) }
+    'hour' ? string with valid_cron_hour(SELF)
+    @documentation{ day of month (1-31) }
+    'day' ? string with valid_cron_day_of_month(SELF)
+    @documentation{ month of year (1-12 or three-letter abbreviated lowercase name) }
+    'month' ? string with valid_cron_month(SELF)
+    @documentation{ day of week (0-7 or three-letter abbreviated lowercase name) }
+    'weekday' ? string with valid_cron_day_of_week(SELF)
+    @documentation{ Interval (in minutes) over which to randomly smear the start time of the job }
     'smear' ? long(0..1440)
 };
 
+@documentation{
+    Validate contents of cron frequency field
+}
+function valid_cron_frequency = {
+    if (ARGC != 1) {
+        error(format('%s: expected 1 parameter, received %d', FUNCTION, ARGC));
+    };
+
+    frequency = to_lowercase(ARGV[0]);
+
+    if (match(frequency, '^(?:@reboot|@yearly|@annually|@monthly|@weekly|@daily|@midnight|@hourly)$')) {
```

Comment:
why the `(?:..)`? also, factor out the leading `@` unless you plan to switch to choice at some point

Reply:
Wrapping things in non-capturing groups is just force of habit, but yet I had thought that this would be an ideal use for `choice`.